File: python/text/text_editor.py
"""Text Editor — buffer model with cursor, edit operations, and undo."""

from dataclasses import dataclass, field
import re


@dataclass
class CursorPos:
    line: int = 0
    col: int = 0


@dataclass
class EditAction:
    """Recorded edit for undo."""
    kind: str  # "insert_char", "delete_char", "insert_line", "delete_line", "replace"
    cursor_before: tuple[int, int] = (0, 0)
    data: dict = field(default_factory=dict)


class TextEditor:
    """A programmatic text buffer with cursor, edit operations, and undo stack."""

    def __init__(self, text: str = ""):
        if text:
            self.lines = text.split("\n")
        else:
            self.lines = [""]
        self.cursor = CursorPos(0, 0)
        self._undo_stack: list[EditAction] = []

    def _clamp_cursor(self):
        self.cursor.line = max(0, min(self.cursor.line, len(self.lines) - 1))
        self.cursor.col = max(0, min(self.cursor.col, len(self.lines[self.cursor.line])))

    def _save(self, kind: str, **data) -> EditAction:
        action = EditAction(
            kind=kind,
            cursor_before=(self.cursor.line, self.cursor.col),
            data=data,
        )
        self._undo_stack.append(action)
        return action

    def insert_char(self, ch: str):
        """Insert character at cursor position, advance cursor."""
        self._save("insert_char", char=ch)
        line = self.lines[self.cursor.line]
        col = self.cursor.col
        self.lines[self.cursor.line] = line[:col] + ch + line[col:]
        self.cursor.col += len(ch)

    def delete_char(self):
        """Backspace: delete character before cursor."""
        if self.cursor.col == 0 and self.cursor.line == 0:
            return
        if self.cursor.col > 0:
            line = self.lines[self.cursor.line]
            deleted = line[self.cursor.col - 1]
            self._save("delete_char", char=deleted)
            self.lines[self.cursor.line] = line[: self.cursor.col - 1] + line[self.cursor.col :]
            self.cursor.col -= 1
        else:
            # Join with previous line
            prev = self.cursor.line - 1
            join_col = len(self.lines[prev])
            self._save("delete_char_join", removed_line=self.lines[self.cursor.line], join_col=join_col)
            self.lines[prev] += self.lines[self.cursor.line]
            del self.lines[self.cursor.line]
            self.cursor.line = prev
            self.cursor.col = join_col

    def insert_line(self):
        """Split current line at cursor (like pressing Enter)."""
        self._save("insert_line")
        line = self.lines[self.cursor.line]
        before = line[: self.cursor.col]
        after = line[self.cursor.col :]
        self.lines[self.cursor.line] = before
        self.lines.insert(self.cursor.line + 1, after)
        self.cursor.line += 1
        self.cursor.col = 0

    def delete_line(self):
        """Remove current line entirely."""
        self._save("delete_line", content=self.lines[self.cursor.line], index=self.cursor.line)
        if len(self.lines) == 1:
            self.lines[0] = ""
            self.cursor.col = 0
        else:
            del self.lines[self.cursor.line]
            if self.cursor.line >= len(self.lines):
                self.cursor.line = len(self.lines) - 1
            self.cursor.col = 0

    def move_cursor(self, direction: str):
        """Move cursor: 'up', 'down', 'left', 'right'."""
        if direction == "left":
            if self.cursor.col > 0:
                self.cursor.col -= 1
            elif self.cursor.line > 0:
                self.cursor.line -= 1
                self.cursor.col = len(self.lines[self.cursor.line])
        elif direction == "right":
            if self.cursor.col < len(self.lines[self.cursor.line]):
                self.cursor.col += 1
            elif self.cursor.line < len(self.lines) - 1:
                self.cursor.line += 1
                self.cursor.col = 0
        elif direction == "up":
            if self.cursor.line > 0:
                self.cursor.line -= 1
                self._clamp_cursor()
        elif direction == "down":
            if self.cursor.line < len(self.lines) - 1:
                self.cursor.line += 1
                self._clamp_cursor()

    def goto(self, line: int, col: int):
        """Move cursor to specific position."""
        self.cursor.line = line
        self.cursor.col = col
        self._clamp_cursor()

    def get_line(self, n: int) -> str:
        """Get content of line n."""
        if 0 <= n < len(self.lines):
            return self.lines[n]
        raise IndexError(f"Line {n} out of range (0..{len(self.lines) - 1})")

    def get_text(self) -> str:
        """Return full buffer content."""
        return "\n".join(self.lines)

    def undo(self):
        """Revert the last edit operation."""
        if not self._undo_stack:
            return
        action = self._undo_stack.pop()
        kind = action.kind
        prev_line, prev_col = action.cursor_before

        if kind == "insert_char":
            ch = action.data["char"]
            line = self.lines[prev_line]
            self.lines[prev_line] = line[:prev_col] + line[prev_col + len(ch):]
        elif kind == "delete_char":
            ch = action.data["char"]
            line = self.lines[prev_line]
            self.lines[prev_line] = line[:prev_col - 1] + ch + line[prev_col - 1:]
        elif kind == "delete_char_join":
            join_col = action.data["join_col"]
            removed = action.data["removed_line"]
            combined = self.lines[prev_line]
            self.lines[prev_line] = combined[:join_col]
            self.lines.insert(prev_line + 1, removed)
        elif kind == "insert_line":
            # Rejoin the two lines
            if prev_line + 1 < len(self.lines):
                self.lines[prev_line] += self.lines[prev_line + 1]
                del self.lines[prev_line + 1]
        elif kind == "delete_line":
            idx = action.data["index"]
            content = action.data["content"]
            if len(self.lines) == 1 and self.lines[0] == "":
                self.lines[0] = content
            else:
                self.lines.insert(idx, content)
        elif kind == "replace":
            self.lines = action.data["old_lines"][:]

        self.cursor.line = prev_line
        self.cursor.col = prev_col

    def find(self, pattern: str) -> list[tuple[int, int]]:
        """Find all occurrences of pattern. Returns list of (line, col)."""
        results = []
        for i, line in enumerate(self.lines):
            start = 0
            while True:
                idx = line.find(pattern, start)
                if idx == -1:
                    break
                results.append((i, idx))
                start = idx + 1
        return results

    def replace(self, pattern: str, replacement: str, count: int = -1) -> int:
        """Replace occurrences of pattern. Returns number of replacements made."""
        self._save("replace", old_lines=self.lines[:])
        total = 0
        for i in range(len(self.lines)):
            if count == -1:
                new_line, n = re.subn(re.escape(pattern), replacement, self.lines[i])
            else:
                remaining = count - total
                if remaining <= 0:
                    break
                new_line, n = re.subn(re.escape(pattern), replacement, self.lines[i], count=remaining)
            if n > 0:
                self.lines[i] = new_line
                total += n
        if total == 0:
            self._undo_stack.pop()  # nothing changed, remove the undo record
        return total


if __name__ == "__main__":
    print("=== Text Editor Demo ===\n")

    ed = TextEditor()

    # Type "Hello, world!"
    for ch in "Hello, world!":
        ed.insert_char(ch)
    print(f"After typing: '{ed.get_text()}'")
    print(f"Cursor: line={ed.cursor.line}, col={ed.cursor.col}")

    # Press Enter and type a second line
    ed.insert_line()
    for ch in "This is line two.":
        ed.insert_char(ch)
    print(f"\nAfter second line:\n{ed.get_text()}")

    # Undo the entire second line
    for _ in range(len("This is line two.")):
        ed.undo()
    ed.undo()  # undo the Enter
    print(f"\nAfter undo back to one line: '{ed.get_text()}'")

    # Restore, add more text
    ed.insert_line()
    for ch in "The quick brown fox jumps.":
        ed.insert_char(ch)
    ed.insert_line()
    for ch in "The fox is quick.":
        ed.insert_char(ch)

    print(f"\nBuffer:\n{ed.get_text()}")

    # Find
    positions = ed.find("fox")
    print(f"\n'fox' found at: {positions}")

    # Replace
    n = ed.replace("fox", "cat")
    print(f"Replaced {n} occurrences of 'fox' -> 'cat'")
    print(f"Buffer:\n{ed.get_text()}")

    # Undo the replace
    ed.undo()
    print(f"\nAfter undo replace:\n{ed.get_text()}")

File: rust/text/text_editor.rs
//! Text Editor — buffer model with cursor, edit operations, and undo.

/// Cursor position in the buffer.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CursorPos {
    pub line: usize,
    pub col: usize,
}

/// A recorded edit action for the undo stack.
#[derive(Debug, Clone)]
enum EditorAction {
    InsertChar {
        cursor: CursorPos,
        ch: char,
    },
    DeleteChar {
        cursor: CursorPos,
        ch: char,
    },
    DeleteCharJoin {
        cursor: CursorPos,
        removed_line: String,
        join_col: usize,
    },
    InsertLine {
        cursor: CursorPos,
    },
    DeleteLine {
        cursor: CursorPos,
        content: String,
        index: usize,
    },
    Replace {
        cursor: CursorPos,
        old_lines: Vec<String>,
    },
}

/// A programmatic text buffer with cursor, edit operations, and undo stack.
pub struct TextEditor {
    lines: Vec<String>,
    cursor: CursorPos,
    undo_stack: Vec<EditorAction>,
}

impl TextEditor {
    /// Create a new editor, optionally initialized with text.
    pub fn new(text: &str) -> Self {
        let lines = if text.is_empty() {
            vec![String::new()]
        } else {
            text.split('\n').map(String::from).collect()
        };
        TextEditor {
            lines,
            cursor: CursorPos { line: 0, col: 0 },
            undo_stack: Vec::new(),
        }
    }

    /// Current cursor position.
    pub fn cursor(&self) -> CursorPos {
        self.cursor
    }

    /// Number of lines in the buffer.
    pub fn line_count(&self) -> usize {
        self.lines.len()
    }

    fn clamp_cursor(&mut self) {
        if self.cursor.line >= self.lines.len() {
            self.cursor.line = self.lines.len() - 1;
        }
        let line_len = self.lines[self.cursor.line].len();
        if self.cursor.col > line_len {
            self.cursor.col = line_len;
        }
    }

    /// Insert a character at the cursor position.
    pub fn insert_char(&mut self, ch: char) {
        self.undo_stack.push(EditorAction::InsertChar {
            cursor: self.cursor,
            ch,
        });
        let line = &mut self.lines[self.cursor.line];
        line.insert(self.cursor.col, ch);
        self.cursor.col += ch.len_utf8();
    }

    /// Delete the character before the cursor (backspace).
    pub fn delete_char(&mut self) {
        if self.cursor.col == 0 && self.cursor.line == 0 {
            return;
        }
        if self.cursor.col > 0 {
            let line = &self.lines[self.cursor.line];
            // Find the char boundary before cursor
            let before = &line[..self.cursor.col];
            let ch = before.chars().last().unwrap();
            let char_len = ch.len_utf8();
            self.undo_stack.push(EditorAction::DeleteChar {
                cursor: self.cursor,
                ch,
            });
            self.lines[self.cursor.line].remove(self.cursor.col - char_len);
            self.cursor.col -= char_len;
        } else {
            // Join with previous line
            let prev = self.cursor.line - 1;
            let join_col = self.lines[prev].len();
            let removed = self.lines.remove(self.cursor.line);
            self.undo_stack.push(EditorAction::DeleteCharJoin {
                cursor: self.cursor,
                removed_line: removed.clone(),
                join_col,
            });
            self.lines[prev].push_str(&removed);
            self.cursor.line = prev;
            self.cursor.col = join_col;
        }
    }

    /// Split the current line at the cursor (Enter key).
    pub fn insert_line(&mut self) {
        self.undo_stack.push(EditorAction::InsertLine {
            cursor: self.cursor,
        });
        let rest = self.lines[self.cursor.line][self.cursor.col..].to_string();
        self.lines[self.cursor.line].truncate(self.cursor.col);
        self.lines.insert(self.cursor.line + 1, rest);
        self.cursor.line += 1;
        self.cursor.col = 0;
    }

    /// Delete the current line entirely.
    pub fn delete_line(&mut self) {
        let content = self.lines[self.cursor.line].clone();
        let index = self.cursor.line;
        self.undo_stack.push(EditorAction::DeleteLine {
            cursor: self.cursor,
            content,
            index,
        });
        if self.lines.len() == 1 {
            self.lines[0].clear();
            self.cursor.col = 0;
        } else {
            self.lines.remove(self.cursor.line);
            if self.cursor.line >= self.lines.len() {
                self.cursor.line = self.lines.len() - 1;
            }
            self.cursor.col = 0;
        }
    }

    /// Move the cursor in a direction: "up", "down", "left", "right".
    pub fn move_cursor(&mut self, direction: &str) {
        match direction {
            "left" => {
                if self.cursor.col > 0 {
                    // Move back one char boundary
                    let before = &self.lines[self.cursor.line][..self.cursor.col];
                    if let Some(ch) = before.chars().last() {
                        self.cursor.col -= ch.len_utf8();
                    }
                } else if self.cursor.line > 0 {
                    self.cursor.line -= 1;
                    self.cursor.col = self.lines[self.cursor.line].len();
                }
            }
            "right" => {
                let line_len = self.lines[self.cursor.line].len();
                if self.cursor.col < line_len {
                    let ch = self.lines[self.cursor.line][self.cursor.col..]
                        .chars()
                        .next()
                        .unwrap();
                    self.cursor.col += ch.len_utf8();
                } else if self.cursor.line < self.lines.len() - 1 {
                    self.cursor.line += 1;
                    self.cursor.col = 0;
                }
            }
            "up" => {
                if self.cursor.line > 0 {
                    self.cursor.line -= 1;
                    self.clamp_cursor();
                }
            }
            "down" => {
                if self.cursor.line < self.lines.len() - 1 {
                    self.cursor.line += 1;
                    self.clamp_cursor();
                }
            }
            _ => {}
        }
    }

    /// Move cursor directly to a position.
    pub fn goto(&mut self, line: usize, col: usize) {
        self.cursor.line = line;
        self.cursor.col = col;
        self.clamp_cursor();
    }

    /// Get the content of line n.
    pub fn get_line(&self, n: usize) -> Option<&str> {
        self.lines.get(n).map(|s| s.as_str())
    }

    /// Get the full buffer content.
    pub fn get_text(&self) -> String {
        self.lines.join("\n")
    }

    /// Undo the last edit operation.
    pub fn undo(&mut self) {
        let action = match self.undo_stack.pop() {
            Some(a) => a,
            None => return,
        };
        match action {
            EditorAction::InsertChar { cursor, ch } => {
                self.lines[cursor.line].remove(cursor.col);
                let _ = ch; // char was at cursor.col
                self.cursor = cursor;
            }
            EditorAction::DeleteChar { cursor, ch } => {
                self.lines[cursor.line].insert(cursor.col - ch.len_utf8(), ch);
                self.cursor = cursor;
            }
            EditorAction::DeleteCharJoin {
                cursor,
                removed_line,
                join_col,
            } => {
                // Split the joined line back apart
                let combined = &self.lines[cursor.line - 1];
                let before = combined[..join_col].to_string();
                let _after = combined[join_col..].to_string();
                self.lines[cursor.line - 1] = before;
                self.lines.insert(cursor.line, removed_line);
                self.cursor = cursor;
            }
            EditorAction::InsertLine { cursor } => {
                // Rejoin the two lines
                if cursor.line + 1 < self.lines.len() {
                    let next = self.lines.remove(cursor.line + 1);
                    self.lines[cursor.line].push_str(&next);
                }
                self.cursor = cursor;
            }
            EditorAction::DeleteLine {
                cursor,
                content,
                index,
            } => {
                if self.lines.len() == 1 && self.lines[0].is_empty() {
                    self.lines[0] = content;
                } else {
                    self.lines.insert(index, content);
                }
                self.cursor = cursor;
            }
            EditorAction::Replace { cursor, old_lines } => {
                self.lines = old_lines;
                self.cursor = cursor;
            }
        }
    }

    /// Find all occurrences of a pattern. Returns vec of (line, col).
    pub fn find(&self, pattern: &str) -> Vec<(usize, usize)> {
        let mut results = Vec::new();
        for (i, line) in self.lines.iter().enumerate() {
            let mut start = 0;
            while let Some(idx) = line[start..].find(pattern) {
                results.push((i, start + idx));
                start += idx + 1;
            }
        }
        results
    }

    /// Replace occurrences of pattern with replacement.
    /// count=-1 replaces all. Returns number of replacements.
    pub fn replace(&mut self, pattern: &str, replacement: &str, count: i32) -> usize {
        self.undo_stack.push(EditorAction::Replace {
            cursor: self.cursor,
            old_lines: self.lines.clone(),
        });
        let mut total = 0usize;
        for i in 0..self.lines.len() {
            if count >= 0 && total >= count as usize {
                break;
            }
            let mut line = self.lines[i].clone();
            let mut new_line = String::new();
            let mut search_start = 0;
            while let Some(idx) = line[search_start..].find(pattern) {
                if count >= 0 && total >= count as usize {
                    new_line.push_str(&line[search_start..]);
                    search_start = line.len();
                    break;
                }
                new_line.push_str(&line[search_start..search_start + idx]);
                new_line.push_str(replacement);
                search_start += idx + pattern.len();
                total += 1;
            }
            new_line.push_str(&line[search_start..]);
            self.lines[i] = new_line;
        }
        if total == 0 {
            self.undo_stack.pop();
        }
        total
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_insert_and_get_text() {
        let mut ed = TextEditor::new("");
        for ch in "hello".chars() {
            ed.insert_char(ch);
        }
        assert_eq!(ed.get_text(), "hello");
        assert_eq!(ed.cursor(), CursorPos { line: 0, col: 5 });
    }

    #[test]
    fn test_insert_line() {
        let mut ed = TextEditor::new("hello world");
        ed.goto(0, 5);
        ed.insert_line();
        assert_eq!(ed.line_count(), 2);
        assert_eq!(ed.get_line(0), Some("hello"));
        assert_eq!(ed.get_line(1), Some(" world"));
    }

    #[test]
    fn test_delete_char() {
        let mut ed = TextEditor::new("abc");
        ed.goto(0, 3);
        ed.delete_char();
        assert_eq!(ed.get_text(), "ab");
    }

    #[test]
    fn test_delete_char_join_lines() {
        let mut ed = TextEditor::new("ab\ncd");
        ed.goto(1, 0);
        ed.delete_char();
        assert_eq!(ed.get_text(), "abcd");
        assert_eq!(ed.cursor(), CursorPos { line: 0, col: 2 });
    }

    #[test]
    fn test_delete_line() {
        let mut ed = TextEditor::new("aaa\nbbb\nccc");
        ed.goto(1, 0);
        ed.delete_line();
        assert_eq!(ed.get_text(), "aaa\nccc");
    }

    #[test]
    fn test_move_cursor() {
        let mut ed = TextEditor::new("abc\ndef");
        ed.goto(0, 1);
        ed.move_cursor("right");
        assert_eq!(ed.cursor(), CursorPos { line: 0, col: 2 });
        ed.move_cursor("down");
        assert_eq!(ed.cursor(), CursorPos { line: 1, col: 2 });
        ed.move_cursor("left");
        assert_eq!(ed.cursor(), CursorPos { line: 1, col: 1 });
        ed.move_cursor("up");
        assert_eq!(ed.cursor(), CursorPos { line: 0, col: 1 });
    }

    #[test]
    fn test_undo_insert_char() {
        let mut ed = TextEditor::new("ab");
        ed.goto(0, 2);
        ed.insert_char('c');
        assert_eq!(ed.get_text(), "abc");
        ed.undo();
        assert_eq!(ed.get_text(), "ab");
    }

    #[test]
    fn test_undo_delete_char() {
        let mut ed = TextEditor::new("abc");
        ed.goto(0, 3);
        ed.delete_char();
        assert_eq!(ed.get_text(), "ab");
        ed.undo();
        assert_eq!(ed.get_text(), "abc");
    }

    #[test]
    fn test_undo_insert_line() {
        let mut ed = TextEditor::new("abcd");
        ed.goto(0, 2);
        ed.insert_line();
        assert_eq!(ed.get_text(), "ab\ncd");
        ed.undo();
        assert_eq!(ed.get_text(), "abcd");
    }

    #[test]
    fn test_undo_delete_line() {
        let mut ed = TextEditor::new("aaa\nbbb");
        ed.goto(0, 0);
        ed.delete_line();
        assert_eq!(ed.get_text(), "bbb");
        ed.undo();
        assert_eq!(ed.get_text(), "aaa\nbbb");
    }

    #[test]
    fn test_find() {
        let ed = TextEditor::new("the cat sat on the mat");
        let results = ed.find("the");
        assert_eq!(results, vec![(0, 0), (0, 15)]);
    }

    #[test]
    fn test_find_multiline() {
        let ed = TextEditor::new("fox here\nno fox there\nanother fox");
        let results = ed.find("fox");
        assert_eq!(results, vec![(0, 0), (1, 3), (2, 8)]);
    }

    #[test]
    fn test_replace_all() {
        let mut ed = TextEditor::new("the fox and the fox");
        let n = ed.replace("fox", "cat", -1);
        assert_eq!(n, 2);
        assert_eq!(ed.get_text(), "the cat and the cat");
    }

    #[test]
    fn test_replace_limited() {
        let mut ed = TextEditor::new("aaa bbb aaa bbb aaa");
        let n = ed.replace("aaa", "xxx", 2);
        assert_eq!(n, 2);
        assert_eq!(ed.get_text(), "xxx bbb xxx bbb aaa");
    }

    #[test]
    fn test_replace_undo() {
        let mut ed = TextEditor::new("foo bar foo");
        ed.replace("foo", "baz", -1);
        assert_eq!(ed.get_text(), "baz bar baz");
        ed.undo();
        assert_eq!(ed.get_text(), "foo bar foo");
    }

    #[test]
    fn test_new_with_multiline_text() {
        let ed = TextEditor::new("line one\nline two\nline three");
        assert_eq!(ed.line_count(), 3);
        assert_eq!(ed.get_line(1), Some("line two"));
    }
}
